### PR TITLE
[usage] Mutex jobs redis WEB-266

### PR DESCRIPTION
**Pull request by easyCZ**
## Description

Uses [redsync](https://github.com/go-redsync/redsync) to provide distributed mutex for scheduled jobs run by the usage component.

* Each job is wrapped in `WithRefreshingMutex` which uses Redis to acquire a mutex
* The mutex has an initial duration, but auto-refreshes if the job takes longer than the initial period

## Related Issue(s)

Fixes #

## How to test

* Unit test
* Preview env has 3 replicas of usage, see logs for some jobs failing due to failing to acquire mutex

## Release Notes

```release-note
NONE
```

## Build Options:

- [ ] /werft with-werft
      Run the build with werft instead of GHA
- [ ] leeway-no-cache
- [ ] /werft no-test
      Run Leeway with `--dont-test`

<details>
<summary>Publish Options</summary>

- [ ] /werft publish-to-npm
- [ ] /werft publish-to-jb-marketplace
</details>

<details>
<summary>Installer Options</summary>

- [ ] with-dedicated-emulation
- [ ] with-ws-manager-mk2
- [ ] workspace-feature-flags
  Add desired feature flags to the end of the line above, space separated
</details>

#### Preview Environment Options:
- [ ] /werft with-local-preview
      If enabled this will build `install/preview`
- [x] /werft with-preview
- [ ] /werft with-large-vm
- [ ] /werft with-gce-vm
      If enabled this will create the environment on GCE infra
- [ ] with-integration-tests=all
      Valid options are `all`, `workspace`, `webapp`, `ide`, `jetbrains`, `vscode`, `ssh`

/hold

### CODE REVIEW

File: components/usage/pkg/scheduler/mutex.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+		}
+	}()
+
+	logger.Debug("Running job inside mutex.")
+	fnErr := fn()
+
+	// release the lock, it will be acquired on subsequent run, possibly by another instance of this job.
+	logger.Debug("Completed job inside mutex. Releasing mutex lock.")
+	if _, err := mutex.UnlockContext(ctx); err != nil {
+		log.Log.WithError(err).Error("Failed to release usage ledger mutex.")
```
Comment (svenefftinge) [inline components/usage/pkg/scheduler/mutex.go:73]: Log message mentions 'ledger'. Also, do you want to use `logger` here (also in goroutine above)?

Comment (easyCZ) [reply components/usage/pkg/scheduler/mutex.go:73]: Yep, will cleanup before opening for review. This is a copy paste from before I generalized the mutex funcation

**Comment by easyCZ** [comment]
Took a long time to open this for review. Had to drop usage of testcontainers as a dependency because they just wouldn't pass in CI (leeway looks to behave different in CI than locally)

File: components/usage/pkg/scheduler/job.go [hunk 2]
```diff
@@ -28,40 +28,8 @@ func NewPeriodicJobSpec(period time.Duration, id string, job Job) (JobSpec, erro
...
 	return JobSpec{
-		Job:      WithoutConcurrentRun(job),
+		Job:      job,
 		ID:       id,
 		Schedule: parsed,
 	}, nil
 }
-
-// WithoutConcurrentRun wraps a Job and ensures the job does not concurrently
-func WithoutConcurrentRun(j Job) Job {
```
Comment (svenefftinge) [inline components/usage/pkg/scheduler/job.go:38]: I liked how you made non concurrent execution a feature of the Job API [here](https://github.com/gitpod-io/gitpod/pull/17433/files#diff-2d1814d4e79d79146e0a3be29f6118d6e93de6bfb60c8f4ce44f93c5a3ccdfbd). Wouldn't it make sense to do it similarly for our go components?

Comment (easyCZ) [reply components/usage/pkg/scheduler/job.go:38]: Sure, I can do that. Originally, the idea was to keep them separate and composable but it looks like in practice we don't actually have a use-case for concurrent runs.

I can refactor it in a follow-up PR to provide a common interface. Landing this as is unblocks running multiple replicas of usage and eventually merging it with usage so would like to land this as is to unblock those

**Review by svenefftinge — APPROVED** [review]

**Pull request merged**